AssertionError: Cannot find .runfiles directory

https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/3319#018ab077-c764-4cc2-8b32-9fb08619ef90

rules_python: rules_python (example bzlmod build file running gazelle and pip integration test on :ubuntu: 20.04 LTS (OpenJDK 11, gcc 9.4.0))
Logs:
```
 File "/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/3b10eb146ec4242f66f1b569edf0069e/execroot/_main/bazel-out/k8-opt-exec-ST-e846b08c7501/bin/external/rules_python_gazelle_plugin~override/python/std_modules", line 172, in FindModuleSpace
    raise AssertionError('Cannot find .runfiles directory for %s' % sys.argv[0])
AssertionError: Cannot find .runfiles directory for /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/3b10eb146ec4242f66f1b569edf0069e/execroot/_main/bazel-out/k8-opt-exec-ST-e846b08c7501/bin/external/rules_python_gazelle_plugin~override/python/std_modules
gazelle: failed to wait for std_modules: exit status 1
```
Culprit Found: [2f9c7d60ea676f9f8e929b20a3ff74e786995c91](https://github.com/bazelbuild/bazel/commit/2f9c7d60ea676f9f8e929b20a3ff74e786995c91)

CC Greenteam @comius

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -58,6 +58,8 @@ A brief description of the categories of changes:
   dependencies improving initial build times involving external dependency
   fetching.
 
+* (gazelle) Improve runfiles lookup hermeticity.
+
 ## [0.25.0] - 2023-08-22
 
 ### Changed

--- a/gazelle/python/parser.go
+++ b/gazelle/python/parser.go
@@ -38,13 +38,20 @@ var (
 )
 
 func startParserProcess(ctx context.Context) {
-	parseScriptRunfile, err := runfiles.Rlocation("rules_python_gazelle_plugin/python/parse")
+	rfiles, err := runfiles.New()
+	if err != nil {
+		log.Printf("failed to create a runfiles object: %v\n", err)
+		os.Exit(1)
+	}
+
+	parseScriptRunfile, err := rfiles.Rlocation("rules_python_gazelle_plugin/python/parse")
 	if err != nil {
 		log.Printf("failed to initialize parser: %v\n", err)
 		os.Exit(1)
 	}
 
 	cmd := exec.CommandContext(ctx, parseScriptRunfile)
+	cmd.Env = rfiles.Env()
 
 	cmd.Stderr = os.Stderr
 

--- a/gazelle/python/parser.go
+++ b/gazelle/python/parser.go
@@ -51,7 +51,7 @@ func startParserProcess(ctx context.Context) {
 	}
 
 	cmd := exec.CommandContext(ctx, parseScriptRunfile)
-	cmd.Env = rfiles.Env()
+	cmd.Env = append(os.Environ(), rfiles.Env()...)
 
 	cmd.Stderr = os.Stderr
 

--- a/gazelle/python/std_modules.go
+++ b/gazelle/python/std_modules.go
@@ -39,7 +39,13 @@ var (
 func startStdModuleProcess(ctx context.Context) {
 	stdModulesSeen = make(map[string]struct{})
 
-	stdModulesScriptRunfile, err := runfiles.Rlocation("rules_python_gazelle_plugin/python/std_modules")
+	rfiles, err := runfiles.New()
+	if err != nil {
+		log.Printf("failed to create a runfiles object: %v\n", err)
+		os.Exit(1)
+	}
+
+	stdModulesScriptRunfile, err := rfiles.Rlocation("rules_python_gazelle_plugin/python/std_modules")
 	if err != nil {
 		log.Printf("failed to initialize std_modules: %v\n", err)
 		os.Exit(1)
@@ -49,7 +55,8 @@ func startStdModuleProcess(ctx context.Context) {
 
 	cmd.Stderr = os.Stderr
 	// All userland site-packages should be ignored.
-	cmd.Env = []string{"PYTHONNOUSERSITE=1"}
+	cmd.Env = append([]string{"PYTHONNOUSERSITE=1"}, rfiles.Env()...)
+
 	stdin, err := cmd.StdinPipe()
 	if err != nil {
 		log.Printf("failed to initialize std_modules: %v\n", err)

